packages/muelu/src/Misc/MueLu_InterfaceAggregationFactory_def.hpp

```diff
@@ -0,0 +1,165 @@
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact
+//                    Jonathan Hu       (jhu@sandia.gov)
+//                    Andrey Prokopenko (aprokop@sandia.gov)
+//                    Ray Tuminaro      (rstumin@sandia.gov)
+//
+// ***********************************************************************
+//
+// @HEADER
+#ifndef MUELU_INTERFACEAGGREGATIONFACTORY_DEF_HPP_
+#define MUELU_INTERFACEAGGREGATIONFACTORY_DEF_HPP_
+
+#include "MueLu_InterfaceAggregationFactory_decl.hpp"
+
+#include "MueLu_Level.hpp"
+#include "MueLu_Aggregates.hpp"
+#include "MueLu_Monitor.hpp"
+
+namespace MueLu
+{
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+InterfaceAggregationFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::InterfaceAggregationFactory()
+{
+}
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+RCP<const ParameterList> InterfaceAggregationFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::GetValidParameterList() const
+{
+  RCP<ParameterList> validParamList = rcp(new ParameterList());
+
+  validParamList->set<RCP<const FactoryBase>>("A", null, "Generating factory of A");
+  validParamList->set<RCP<const FactoryBase>>("Aggregates", null, "Generating factory of the Aggregates (for block 0,0)");
+
+  validParamList->set<LocalOrdinal>("number of DOFs per node", Teuchos::ScalarTraits<LocalOrdinal>::one(), "Number of DOFs per node");
+
+  return validParamList;
+}
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void InterfaceAggregationFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::DeclareInput(Level &currentLevel) const
+{
+  Input(currentLevel, "A");
+  Input(currentLevel, "Aggregates");
+
+  currentLevel.DeclareInput("Lagr2Dof", NoFactory::get(), this);
```

Comment:
This is a simple and maybe "hacky" way to get this information. Fine for me, but maybe in between there is a smarter way to do so (e.g., getting this information from a parameter list on the finest level). Not that i'm aware of it though.

Reply:
If other people want to use this factory, they have to be aware how to do so. Maybe one should add a comment or point to the unit test or add a small example in the tutorial.